### PR TITLE
Fix compile support for CTC loss tensor lengths

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184573

Teach FakeTensor and meta handling for CTC loss selector/internal paths while capturing the public tensor-length CTC op for Dynamo/Inductor fallback, so torch.compile preserves runtime backend selection.\n\nFixes #183396\nGenerated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by ezyang** [comment]
@claude Do a side-by-side comparison of the C++ CTC loss code and the new meta implementation. Do we have 100% fidelity. Note that you have to do both CPU and CUDA

**Review by ezyang — CHANGES_REQUESTED** [review]
It shouldn't be necessary to special case the Dynamo code

**Comment by jansel** [comment]
My agent says I addressed the Dynamo special-case review by removing the torch.ctc_loss/F.ctc_loss handler from Dynamo and moving the support to CTC fake/meta base-operator handling plus the existing public-op Inductor fallback. The updated graph test now verifies the public CTC call stays in the graph and the internal selector/_ctc_loss ops do not leak into Dynamo.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I inspected the Dr. CI new failures and they were checkout/setup failures before tests ran: git promisor fetch/auth errors, followed by missing workflow helper scripts because checkout did not complete. I reran the failed jobs for the pull, inductor, and torchtitan workflows and will revisit once the reruns report back.

**Comment by ezyang** [comment]
@claude How come `make_fallback(aten.ctc_loss.Tensor, warn=False, override_decomp=True)` ? Why overriding decomp?

**Comment by ezyang** [comment]
@claude How come we need both fake and meta impls?

File: torch/_subclasses/fake_impls.py [hunk 1]
```diff
@@ -363,6 +363,179 @@ def _sparse_coo_tensor_with_dims_and_tensors(
...
+
+    if len(input_lengths) != len(target_lengths):
+        return False
+    max_input_length = log_probs.size(0)
+    return all(
+        input_length == max_input_length for input_length in input_lengths
+    ) and all(
+        target_length < 256 and target_length <= input_lengths[i]
+        for i, target_length in enumerate(target_lengths)
+    )
```
Comment (ezyang) [inline torch/_subclasses/fake_impls.py]: Truly deranged

**Comment by ezyang** [comment]
@claude

This is deeply sus

```
def _use_ctc_loss_backend_tensor(
    fake_mode: FakeTensorMode,
    func: OpOverload,
    log_probs: FakeTensorLike,
    targets: FakeTensorLike,
    input_lengths: FakeTensorLike,
    target_lengths: FakeTensorLike,
    blank: int,
) -> bool:
    # Tensor length values are unavailable under FakeTensor. Use the
    # shape-equivalent generic CTC path for metadata; runtime public CTC
    # dispatch still performs backend selection with real length values.
    return False
```

If it is the case that the choice here cannot influence the final stride choice, then we should have refactored the code not to query this at all.

**Comment by jansel** [comment]
@claude review these changes

File: torch/_subclasses/fake_impls.py (hunk 1 above)
Comment (jansel) [reply torch/_subclasses/fake_impls.py]: My agent says I refactored this so FakeTensor no longer answers the internal cuDNN/MIOpen backend selector with a hard-coded False. Tensor-subclass CTC loss now takes the generic _ctc_loss path before backend selector queries, and fake support is limited to _ctc_loss.Tensor metadata for the hidden log_alpha shape. The graph regression still verifies the internal selector ops do not appear.

**Comment by jansel** [comment]
My agent says I addressed the current Dr. CI failures on the PR. The C++ build failure was caused by ctc_loss_tensor_subclass having external linkage without a prototype, so it now stays in the existing anonymous namespace. The export-opinfo failure was an unexpected success, so I removed the stale nn.functional.ctc_loss xfail. I validated with the targeted export-opinfo test, the CTC FakeTensor tests, an incremental LossCTC.cpp object build, and lintrunner -a.

**Comment by jansel** [comment]
My agent says I confirmed the June 5 concern about returning False from _use_ctc_loss_backend_tensor was addressed by the follow-up refactor: tensor-subclass CTC loss now avoids querying the cuDNN/MIOpen backend selector during fake metadata handling and takes the generic _ctc_loss path, so the fake path no longer bakes in a backend-selection answer.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
